skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+        boundary point of a interval feature is unknown. The location may
+        begin or end at some points outside the specified coordinates. This
+        accommodates the location format [1]_ of INSDC.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as "strand", "gene_name", or "product".
+
+    Attributes
+    ----------
+    locations
+    boundaries
+    metadata
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+
+    Examples
+    --------
+    >>> from skbio.metadata import Interval, IntervalMetadata
+    >>> gene = Interval(interval_metadata=IntervalMetadata(),
+    ...                 locations=[(1, 2), (4, 7)],
+    ...                 metadata={'name': 'sagA'})
+    >>> gene    # doctest: +ELLIPSIS
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'name': 'sagA'})
+
+    References
+    ----------
+    .. [1] ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3
+    """
+    def __init__(self, interval_metadata, locations,
+                 boundaries=None, metadata=None):
+        if not isinstance(interval_metadata, IntervalMetadata):
+            raise TypeError('You need to provide an IntervalMetadata'
+                            'object, not %r' % interval_metadata)
+        # Intervals
+        self._interval_metadata = interval_metadata
+
+        locations = list(locations)
+        self.locations = locations
+
+        # Boundaries: has to be after locations assignment because
+        # locations setter reset boundaries.
+        indices = [i[0] for i in sorted(enumerate(locations),
+                                        key=operator.itemgetter(1))]
+        # use the boundaries setter to validate the value
+        self.boundaries = boundaries
+        self.boundaries = [self.boundaries[i] for i in indices]
+
+        # Metadata
+        if metadata is None:
+            metadata = {}
+        self.metadata = metadata
+
+        # add this interval feature to the associated IntervalMetadata
+        self._add()
+
+    def _add(self):
+        """Add the current ``Interval`` to the IntervalMetadata object."""
+        for loc in self.locations:
+            start, end = loc
+            self._interval_metadata._interval_tree.add(start, end, self)
+        self._interval_metadata._intervals.append(self)
+
+    @experimental(as_of='0.5.0-dev')
+    def __eq__(self, other):
+        '''Test if this ``Interval`` object is equal to another.
+
+        The equality is performed by checking if the ``metadata``,
+        ``location`` and ``boundaries`` are equal.
+
+        Parameters
+        ----------
+        other : Interval
+            Interval to test for equality against.
+
+        Returns
+        -------
+        bool
+            Indicates if the two objects are equal.
+        '''
+        return ((self.metadata == other.metadata) and
```

Comment:
Can you add a comment explaining that this is deep equality because `self.locations` and `self.boundaries` are always sorted and so permutations during construction won't matter to equality? I almost commented about that until I realized they were sorted.